libnetdata/popen/popen.c

```diff
@@ -2,81 +2,129 @@
-    if (myp_tracking == 0)
+static void netdata_popen_tracking_lock(void) {
+    if(!netdata_popen_tracking_enabled)
         return;
 
-    netdata_mutex_lock(&myp_lock);
+    netdata_mutex_lock(&netdata_popen_tracking_mutex);
 }
 
 // myp_add_unlock release the lock if we're tracking.
-static void myp_add_unlock(void) {
-    if (myp_tracking == 0)
+static void netdata_popen_tracking_unlock(void) {
+    if(!netdata_popen_tracking_enabled)
         return;
 
-    netdata_mutex_unlock(&myp_lock);
+    netdata_mutex_unlock(&netdata_popen_tracking_mutex);
 }
 
 // myp_add_locked adds pid if we're tracking.
 // myp_add_lock must have been called previously.
-static void myp_add_locked(pid_t pid) {
-    struct mypopen *mp;
-
-    if (myp_tracking == 0)
+static void netdata_popen_tracking_add_pid_unsafe(pid_t pid) {
+    if(!netdata_popen_tracking_enabled)
         return;
 
-    mp = mallocz(sizeof(struct mypopen));
+    struct netdata_popen *mp;
+
+    mp = mallocz(sizeof(struct netdata_popen));
     mp->pid = pid;
 
-    mp->next = mypopen_root;
-    mp->prev = NULL;
-    if (mypopen_root != NULL)
-        mypopen_root->prev = mp;
-    mypopen_root = mp;
-    netdata_mutex_unlock(&myp_lock);
+    DOUBLE_LINKED_LIST_PREPEND_UNSAFE(netdata_popen_root, mp, prev, next);
 }
 
 // myp_del deletes pid if we're tracking.
-static void myp_del(pid_t pid) {
-    struct mypopen *mp;
-
-    if (myp_tracking == 0)
+static void netdata_popen_tracking_del_pid(pid_t pid) {
+    if(!netdata_popen_tracking_enabled)
         return;
 
-    netdata_mutex_lock(&myp_lock);
-    for (mp = mypopen_root; mp != NULL; mp = mp->next) {
-        if (mp->pid == pid) {
-            if (mp->next != NULL)
-                mp->next->prev = mp->prev;
-            if (mp->prev != NULL)
-                mp->prev->next = mp->next;
-            if (mypopen_root == mp)
-                mypopen_root = mp->next;
-            freez(mp);
+    struct netdata_popen *mp;
+
+    netdata_mutex_lock(&netdata_popen_tracking_mutex);
+
+    DOUBLE_LINKED_LIST_FOREACH_FORWARD(netdata_popen_root, mp, prev, next) {
+        if(unlikely(mp->pid == pid))
             break;
-        }
     }
 
-    if (mp == NULL)
+    if(mp) {
+        DOUBLE_LINKED_LIST_REMOVE_UNSAFE(netdata_popen_root, mp, prev, next);
+        freez(mp);
+    }
+    else
         error("Cannot find pid %d.", pid);
 
-    netdata_mutex_unlock(&myp_lock);
+    netdata_mutex_unlock(&netdata_popen_tracking_mutex);
 }
 
-#define PIPE_READ 0
-#define PIPE_WRITE 1
+// netdata_popen_tracking_init() should be called by apps which act as init
+// (pid 1) so that processes created by mypopen and mypopene
+// are tracked. This enables the reaper to ignore processes
+// which will be handled internally, by calling myp_reap, to
+// avoid issues with already reaped processes during wait calls.
+//
+// Callers should call myp_free() to clean up resources.
+void netdata_popen_tracking_init(void) {
+    info("process tracking enabled.");
```

Comment:
For these lines we have:

```sh
*** CID 357339:  Concurrent data access violations  (MISSING_LOCK)
/libnetdata/popen/popen.c: 82 in netdata_popen_tracking_init()
76     // which will be handled internally, by calling myp_reap, to
77     // avoid issues with already reaped processes during wait calls.
78     //
79     // Callers should call myp_free() to clean up resources.
80     void netdata_popen_tracking_init(void) {
81         info("process tracking enabled.");
>>>     CID 357339:  Concurrent data access violations  (MISSING_LOCK)
>>>     Accessing "netdata_popen_tracking_enabled" without holding lock "netdata_popen_tracking_mutex". Elsewhere, "netdata_popen_tracking_enabled" is accessed with "netdata_popen_tracking_mutex" held 1 out of 2 times (1 of these accesses strongly imply that it is necessary).
82         netdata_popen_tracking_enabled = true;
83     
84         if (netdata_mutex_init(&netdata_popen_tracking_mutex) != 0)
85             fatal("netdata_popen_tracking_init() mutex init failed.");
86     }
87     

** CID 357338:  Resource leaks  (RESOURCE_LEAK)
/streaming/receiver.c: 710 in rrdpush_receive()
```

Reply:
This seems a false positive - it is the init function - the mutex is not initialized yet.